meson.build

```diff
@@ -0,0 +1,1207 @@
+
+sip_autoconf = configuration_data()
+configure_file(output : 'pjsip_sip_autoconf.h', configuration : sip_autoconf)
+run_command('/bin/cp', 'builddir/pjsip_sip_autoconf.h', 'pjsip/include/pjsip/sip_autoconf.h', check : true)
+
+#
+
+add_global_arguments('-fstack-protector-strong', language : 'c')
+add_global_arguments('-Wformat', language : 'c')
+add_global_arguments('-Werror=format-security', language : 'c')
+add_global_arguments('-ggdb3', language : 'c')
+
+add_global_arguments('-DPJ_AUTOCONF=1', language : ['c', 'cpp', 'objc'])
+
+add_global_arguments('-DPJ_SOCK_HAS_GETADDRINFO=1', language : ['c', 'cpp'])
+add_global_arguments('-DPJ_HAS_IPV6=0', language : ['c', 'cpp'])
+add_global_arguments('-DPJ_SOCK_HAS_INET_PTON=0', language : ['c', 'cpp'])
+add_global_arguments('-DPJ_GETHOSTIP_DISABLE_LOCAL_RESOLUTION=1', language : ['c', 'cpp'])
+
+add_global_arguments('-DPJMEDIA_HAS_WEBRTC_AEC=0', language : ['c', 'cpp'])
+add_global_arguments('-DPJMEDIA_HAS_WEBRTC_AEC3=0', language : ['c', 'cpp'])
+
+add_global_arguments('-DPJMEDIA_AUDIO_DEV_HAS_WMME=0', language : ['c', 'cpp'])
+add_global_arguments('-DPJMEDIA_AUDIO_DEV_HAS_PORTAUDIO=0', language : ['c', 'cpp'])
+add_global_arguments('-DPJMEDIA_AUDIO_DEV_HAS_WMME=0', language : ['c', 'cpp'])
+
+add_global_arguments('-DPJMEDIA_HAS_LIBAVFORMAT=0', language : ['c', 'cpp'])
+add_global_arguments('-DPJMEDIA_HAS_LIBAVCODEC=0', language : ['c', 'cpp'])
+add_global_arguments('-DPJMEDIA_HAS_LIBSWSCALE=0', language : ['c', 'cpp'])
+add_global_arguments('-DPJMEDIA_HAS_LIBAVUTIL=0', language : ['c', 'cpp'])
+
+add_global_arguments('-DPJMEDIA_HAS_ILBC_CODEC=0', language : ['c', 'cpp'])
+add_global_arguments('-DPJMEDIA_HAS_GSM_CODEC=0', language : ['c', 'cpp'])
+add_global_arguments('-DPJMEDIA_HAS_L16_CODEC=0', language : ['c', 'cpp'])
+add_global_arguments('-DPJMEDIA_HAS_G722_CODEC=0', language : ['c', 'cpp'])
+
+system_dep = []
+
+openssl_dep = dependency('openssl')
+
+thread_dep = dependency('threads', required : true)
+
+uuid_dep = dependency('uuid', required : true)
+
+cc = meson.get_compiler('c')
+m_dep = cc.find_library('m', required : true)
+
+srtp = get_option('srtp')
+if srtp == 'system'
+  srtp2_dep = dependency('libsrtp2', required : true)
+  add_global_arguments('-DPJMEDIA_HAS_SRTP=1', language : ['c', 'cpp'])
+  add_global_arguments('-DPJMEDIA_EXTERNAL_SRTP=2', language : ['c', 'cpp'])
+elif srtp == 'external'
+  # TODO - use https://github.com/cisco/libsrtp/blob/main/meson.build
+elif srtp == 'bundled'
+  # TODO - last chance use bundled version... 
+endif
+
+speex = get_option('speex')
+if speex == 'system'
+  speex_dep = dependency('speex', required : true)
+  add_global_arguments('-DPJMEDIA_HAS_SPEEX_AEC=0', language : ['c', 'cpp'])
+  add_global_arguments('-DPJMEDIA_HAS_SPEEX_CODEC=0', language : ['c', 'cpp'])
+elif speex == 'external'
+  # TODO - use https://gitlab.xiph.org/xiph/speex/-/blob/master/meson.build
+elif speex == 'bundled'
+  # TODO - last chance use bundled version...
+endif
```

Comment:
These options seem entirely too complex, and meson can handle this automatically via `-Dwrap_mode=forcefallback` etc to force using the bundled copy instead of the system one. I don't see the need to reinvent the significant effort that the meson developers have put into making this polished.

Other build systems don't have an integrated way of handling this well, so third_party directories with hardcoded vendor copies are used as a hack.

Reply:
I don't quite understand what exactly you are suggesting. Now there is some directory structure and source code in PJPROEJCT. Which other developers use since 2005, so I still don't understand how it is possible to support `third_party`, external source and using the system library without significant tree changes. Please tell me your idea in more detail.

Reply:
I guess the issue is that it's not entirely clear what the goal of each option is.

- The system option has an obvious purpose -- many people like building against system shared libraries for consistency and applying security fixes once globally, e.g. buildroot would want to do this
- The external option would use a meson wrap, I guess, which is useful if you don't have a system copy available or you want to build statically linked executables for deployment or just want to test against whichever versions were tested as wraps
- The bundled option would *not* use a meson wrap, but it would... use the third_party/ directory and you'd write your own meson.build files for each one?

What is the difference between options 2 and 3? When would one prefer 2 over 3, and when would one prefer 3 over 2?

Reply:
> What is the difference between options 2 and 3? When would one prefer 2 over 3, and when would one prefer 3 over 2?

Now the architecture of the project is exactly like this and I think that there are some good reasons for this. In particular, I think that the goal is to ensure the autonomy of the assembly. In any case, I don't think it's in my power to make any structure changes.

I think you may wait for the first integration of meson.build and open discussions and make suggestions. Now I make reference version, with demonstration of options for solving the problem with dependencies.